### PR TITLE
Fixes #3183 - Ensure timestamp fact is a string, not a symbol

### PR DESCRIPTION
**Pull request by GregSutcliffe**
Do we need the db:migration? I can't find a `:_timestamp` fact_name in any of my DBs...

### CODE REVIEW

File: app/models/host/base.rb [hunk 1]
```diff
@@ -70,6 +70,12 @@ def importFacts facts
 
       # we are not importing facts for hosts in build state (e.g. waiting for a re-installation)
       raise ::Foreman::Exception.new("Host is pending for Build") if build
+
+      # Ensure timestamp is always a string
+      if facts.include?(:_timestamp)
+        timestamp = facts.delete(:_timestamp)
+        facts['_timestamp'] = timestamp unless facts['_timestamp'].present?
```
Comment (ohadlevy) [inline app/models/host/base.rb:77]: ||= ?

**Comment by ohadlevy** [comment]
in theory, for hosts did not continue to send newer facts, they might still have the :_timestamp fact in the db, I would create a migration to find and rename those?

**Comment by domcleal** [comment]
I don't understand what problem this solves, is it hypothetical, or do we have an example database with this in it?

The scopes appear to work correctly, they convert the symbol to a string and search on that, rather than keeping the ":" prefix or anything (on develop):

<pre>
2.0.0p247 :020 >   FactValue.timestamp_facts
  FactValue Load (0.1ms)  SELECT "fact_values".* FROM "fact_values" INNER JOIN "fact_names" ON "fact_names"."id" = "fact_values"."fact_name_id" WHERE (fact_names.name = '_timestamp')
 => [] 
</pre>

File: app/models/host/base.rb [hunk 2]
```diff
@@ -70,6 +70,12 @@ def importFacts facts
 
       # we are not importing facts for hosts in build state (e.g. waiting for a re-installation)
       raise ::Foreman::Exception.new("Host is pending for Build") if build
+
+      # Ensure timestamp is always a string
+      if facts.include?(:_timestamp)
```
Comment (domcleal) [inline app/models/host/base.rb:75]: Given we access this on line 79 as a symbol and set it with a string on line 77, the hash must be with indifferent access anyway.. so what does this solve?

**Comment by ohadlevy** [comment]
@domcleal my only concern is that if a user will not import new facts, the old time stamp fact would still be in the db, you are correct that once the the old name is no longer in the db, thats invalid, hence why I suggested to do it in a migration.

**Comment by domcleal** [comment]
@ohadlevy what old versus new fact in the DB?  We don't store `_timestamp` with a ":" prefix today, do we?  It's a symbol that'll get mapped to a string and always stored as `_timestamp`.

**Comment by ohadlevy** [comment]
nope.. in v1.2 and older it was '<rubysym>_timestamp' in the db.

**Comment by domcleal** [comment]
Not in my 1.2.3 instance with PostgreSQL:

```
48 | _timestamp             | 2013-10-07 11:55:46.713036 | 2013-10-07 11:55:46.713036
```

**Comment by ohadlevy** [comment]
Interesting.. on my sqlite I can see it.

**Comment by GregSutcliffe** [comment]
I can't reproduce it either, I wrote the this "blind" as none of my databases contain `<rubysym>_timestamp`. @ohadlevy could it just be your db? :P

**Comment by ohadlevy** [comment]
OK, I give up then, it was probably in an older rails version of something.

**Comment by GregSutcliffe** [comment]
We can always revisit this if it turns up in the wild.

**Pull request closed without merging**